### PR TITLE
ports Radial Menu for service borg + RSF improvements from TG

### PR DESCRIPTION
**Pull request by Jumps0**
### Intent of your Pull Request

This combines two PR's from TG, those being:
RSF PR but again because github is **** #50646 by **LemonInTheDark**
Mining and Service borgs now use a radial menu to choose their skins #50310 by **Arkatos1**

Credit to the respective coders, I just ported it here.

This PR gives the service borg a radial select menu for its skins instead of a list menu. It also improves the RSF by giving it a skin (from TG) instead of just using the RCD one. It also makes the RSF code better, to quote the original PR "Cleans up rsf code to make it more modular, allows for sublists in arrays, ect."

Images:
Screenshot (349)
Screenshot (351)
rsf

#### Changelog

:cl:  LemonInTheDark & Arkatos
rscadd: Both Service borg skin select & RSF radial menus
tweak: RSF's are now modular
tweak: Subtypes of RSFs can now have sublists
imageadd: RSF icons from TG
/:cl:

### CODE REVIEW

**Comment by Jumps0** [comment]
this is so sad

**Comment by Jumps0** [comment]
sorry Ktl, i added like 1 word

File: code/game/objects/items/RSF.dm [hunk 1]
```diff
@@ -3,183 +3,203 @@ CONTAINS:
...
+	var/spent_icon_state = "rsf_empty"
 	lefthand_file = 'icons/mob/inhands/equipment/tools_lefthand.dmi'
 	righthand_file = 'icons/mob/inhands/equipment/tools_righthand.dmi'
 	opacity = 0
 	density = FALSE
 	anchored = FALSE
 	item_flags = NOBLUDGEON
 	armor = list("melee" = 0, "bullet" = 0, "laser" = 0, "energy" = 0, "bomb" = 0, "bio" = 0, "rad" = 0, "fire" = 0, "acid" = 0)
+	///The current matter count
 	var/matter = 0
```
Comment (Hopekz) [inline code/game/objects/items/RSF.dm]:
```suggestion
	var/matter = 0 // The current matter count
```

File: code/game/objects/items/RSF.dm [hunk 2]
```diff
@@ -3,183 +3,203 @@ CONTAINS:
...
+///Extracts the related object from a associated list of objects and values, or lists and objects.
+#define OBJECT_OR_LIST_ELEMENT(from, input) (islist(input) ? from[input] : input)
 /obj/item/rsf
 	name = "\improper Rapid-Service-Fabricator"
 	desc = "A device used to rapidly deploy service items."
 	icon = 'icons/obj/tools.dmi'
-	icon_state = "rcd"
+	icon_state = "rsf" ///thanks tg
+	///The icon state to revert to when the tool is empty
+	var/spent_icon_state = "rsf_empty"
```
Comment (Hopekz) [inline code/game/objects/items/RSF.dm]:
```suggestion
	var/spent_icon_state = "rsf_empty" // The icon state to revert to when the tool is empty
```

File: code/game/objects/items/RSF.dm [hunk 3]
```diff
@@ -3,183 +3,203 @@ CONTAINS:
...
 	armor = list("melee" = 0, "bullet" = 0, "laser" = 0, "energy" = 0, "bomb" = 0, "bio" = 0, "rad" = 0, "fire" = 0, "acid" = 0)
+	///The current matter count
 	var/matter = 0
-	var/mode = 1
+	///The max amount of matter in the device
+	var/max_matter = 30
+	///The type of the object we are going to dispense
+	var/to_dispense
+	///The cost of the object we are going to dispense
+	var/dispense_cost = 0
```
Comment (Hopekz) [inline code/game/objects/items/RSF.dm]:
```suggestion
	var/max_matter = 30 //The max amount of matter in the device
	var/to_dispense //The type of the object we are going to dispense
	var/dispense_cost = 0 // The cost of the object we are going to dispense
```

File: code/game/objects/items/RSF.dm [hunk 4]
```diff
@@ -3,183 +3,203 @@ CONTAINS:
...
+								/obj/item/clothing/mask/cigarette = 10,
+								)
+	///An associated list of fuel and it's value
+	var/list/matter_by_item = list(/obj/item/rcd_ammo = 10,)
+	///A list of surfaces that we are allowed to place things on.
+	var/list/allowed_surfaces = list(/turf/open/floor, /obj/structure/table)
+	///The unit of mesure of the matter, for use in text
+	var/discriptor = "fabrication-units"
+	///The verb that describes what we're doing, for use in text
+	var/action_type = "Dispensing"
```
Comment (Hopekz) [inline code/game/objects/items/RSF.dm]:
```suggestion
	var/list/matter_by_item = list(/obj/item/rcd_ammo = 10,) // An associated list of fuel and it's value
	var/list/allowed_surfaces = list(/turf/open/floor, /obj/structure/table) // A list of surfaces that we are allowed to place things on.
	var/discriptor = "fabrication-units" // The unit of mesure of the matter, for use in text
	var/action_type = "Dispensing" // The verb that describes what we're doing, for use in text
```

File: code/game/objects/items/RSF.dm [hunk 5]
```diff
@@ -3,183 +3,203 @@ CONTAINS:
...
+	//Warning, prepare for bodgecode
+	while(islist(target))//While target is a list we continue the loop
+		var/picked = show_radial_menu(user, src, formRadial(target), custom_check = CALLBACK(src, .proc/check_menu, user), require_near = TRUE)
+		if(!check_menu(user) || picked == null)
+			return
+		for(var/emem in target)//Back through target agian
+			var/atom/test = OBJECT_OR_LIST_ELEMENT(target, emem)
+			if(picked == initial(test.name))//We try and find the entry that matches the radial selection
+				cost = target[emem]//We cash the cost
+				target = emem
```
Comment (Hopekz) [inline code/game/objects/items/RSF.dm:77]: How come TG does their comments wacky but then does it our way here I cry

File: code/game/objects/items/RSF.dm [hunk 6]
```diff
@@ -3,183 +3,203 @@ CONTAINS:
...
+				break
+		//If we found a list we start it all again, this time looking through its contents.
+		//This allows for sublists
+	to_dispense = target
```
Comment (Hopekz) [inline code/game/objects/items/RSF.dm]:
```suggestion
	to_dispense = target //This allows for sublists
```

File: code/game/objects/items/RSF.dm [hunk 7]
```diff
@@ -3,183 +3,203 @@ CONTAINS:
...
+	cost_by_item = list(/obj/item/reagent_containers/food/snacks/cookie = 100)
+	dispense_cost = 100
+	discriptor = "cookie-units"
+	action_type = "Fabricates"
+	///Tracks whether or not the cookiesynth is about to print a poisoned cookie
+	var/toxin = FALSE //This might be better suited to some initialize fuckery, but I don't have a good "poisoned" sprite
+	///Holds a copy of world.time taken the last time the synth gained a charge. Used with cooldowndelay to track when the next charge should be gained
 	var/cooldown = 0
+	///The period between recharges
 	var/cooldowndelay = 10
```
Comment (Hopekz) [inline code/game/objects/items/RSF.dm]:
```suggestion
	var/cooldown = 0 //Holds a copy of world.time taken the last time the synth gained a charge. Used with cooldowndelay to track when the next charge should be gained
	var/cooldowndelay = 10 //The period between recharges
```

File: code/modules/mob/living/silicon/robot/robot_modules.dm [hunk 8]
```diff
@@ -232,6 +232,19 @@
 		R.hud_used.update_robot_modules_display()
 	SSblackbox.record_feedback("tally", "cyborg_modules", 1, R.module)
 
+   /*
+   check_menu: Checks if we are allowed to interact with a radial menu
+  
+   Arguments:
+   user The mob interacting with a menu
+  */
```
Comment (Hopekz) [inline code/modules/mob/living/silicon/robot/robot_modules.dm:240]: So super pretty!

**Review by Hopekz — CHANGES_REQUESTED** [review]
AAAAAH

File: code/game/objects/items/RSF.dm (hunk 5 above)
Comment (Bibby0110) [reply code/game/objects/items/RSF.dm:77]: Because this isn't autodocced, this is a regular comment.

**Review by Hopekz — APPROVED** [review]

**Pull request merged**